src/semidiscretization/semidiscretization_coupled.jl

```diff
@@ -0,0 +1,366 @@
+
+    integrate(func, u, mesh, equations, solver, cache, normalize=false)
+  end
+
+  # Normalize with total volume
+  if normalize
+    total_volume_ = total_volume(semi)
+    integral = integral / total_volume_
+  end
+
+  return integral
+end
+
+
+function total_volume(semi::SemidiscretizationCoupled)
+  sum(semi.semis) do semi_
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semi_)
+    total_volume(mesh, solver, cache)
+  end
+end
+
+
+function compute_coefficients(t, semi::SemidiscretizationCoupled)
+  @unpack u_indices = semi
+
+  u_ode = Vector{real(semi)}(undef, u_indices[end][end])
+
+  for i in 1:nmeshes(semi)
+    # Call `compute_coefficients` in `src/semidiscretization/semidiscretization.jl`
+    u_ode[u_indices[i]] .= compute_coefficients(t, semi.semis[i])
+  end
+
+  return u_ode
+end
+
+
+function allocate_coupled_boundary_conditions(boundary_conditions, semi) end
+
+function allocate_coupled_boundary_conditions(boundary_conditions::Union{Tuple, NamedTuple}, semi)
+  n_boundaries = 2 * ndims(semi)
+  mesh, equations, solver, _ = mesh_equations_solver_cache(semi)
+
+  for direction in 1:n_boundaries
+    boundary_condition = semi.boundary_conditions[direction]
+
+    allocate_coupled_boundary_condition(boundary_condition, direction, mesh, equations, solver)
+  end
+end
+
+function allocate_coupled_boundary_condition(boundary_condition, direction, mesh, equations, solver) end
+
+# In 2D
+function allocate_coupled_boundary_condition(boundary_condition::BoundaryConditionCoupled{3}, direction, mesh, equations, dg::DG)
+  if direction in (1, 2)
+    cell_size = size(mesh, 2)
+  else
+    cell_size = size(mesh, 1)
+  end
+  boundary_condition.u_boundary = Array{Float64, 3}(undef, nvariables(equations), nnodes(dg), cell_size)
+end
+
+# In 3D
+function allocate_coupled_boundary_condition(boundary_condition::BoundaryConditionCoupled{5}, direction, mesh, equations, dg::DG)
+  if direction in (1, 2)
+    cell_size = (size(mesh, 2), size(mesh, 3))
+  elseif direction in (3, 4)
+    cell_size = (size(mesh, 1), size(mesh, 3))
+  else # direction in (5, 6)
+    (size(mesh, 1), size(mesh, 2))
+  end
+  boundary_condition.u_boundary = Array{Float64, 5}(undef, nvariables(equations), nnodes(dg), nnodes(dg), cell_size...)
+end
+
+
+function rhs!(du_ode, u_ode, semi::SemidiscretizationCoupled, t)
+  @unpack u_indices = semi
+
+  time_start = time_ns()
+
+  @trixi_timeit timer() "copy to coupled boundaries" begin
+    for semi_ in semi.semis
+      copy_to_coupled_boundary(semi_.boundary_conditions, u_ode, semi)
+    end
+  end
+
+  # Call rhs! for each semidiscretization
+  for i in 1:nmeshes(semi)
+    u_loc  = @view u_ode[u_indices[i]]
+    du_loc = @view du_ode[u_indices[i]]
+
+    rhs!(du_loc, u_loc, semi.semis[i], t)
+  end
+
+  runtime = time_ns() - time_start
+  put!(semi.performance_counter, runtime)
+
+  return nothing
+end
+
+
+function copy_to_coupled_boundary(boundary_condition, u_ode, semi) end
```

Comment:
Why this definition (I think there's a single answer to all these questions, so feel free to only answer the first - this is just such that I don't forget about it)

Reply:
See above